Edits to storage topic and light touch on supported releases (#3634)
cloudnative-pg/cloudnative-pg: docs/src/storage.md

solution, especially if you manage very large databases.

!!! Important
    See the Kubernetes documentation for a list of all
    the supported [container storage interface (CSI) drivers](https://kubernetes-csi.github.io/docs/drivers.html)
    that provide snapshot capabilities.

## Benchmarking CloudNativePG

Before deploying the database in production, we recommend that you benchmark 
CloudNativePG in a controlled Kubernetes environment. Follow
the guidelines in [Benchmarking](benchmarking.md).

Briefly, we recommend operating at two levels:

> - Measuring the performance of the underlying storage using fio, with relevant

Author: We do not use tagging on names of tools, but we do on names of commands. In this case, I think you mean the tool. The fio website doesn't use any special tagging. So If you want to change it to "the `fio` command, the backticks are okay. Otherwise, please do not include the backticks.

  metrics for database workloads such as throughput for sequential reads, sequential

>   writes, random reads, and random writes

Author: In this case, it's a list of items, none of which are sentences. 

If it were a definition list (which also might have phrases and not sentences), I try to always use this form:

**Thing being defined** em-dash The definition, ending with a period.

I am not 100% sure all of this thinking is captured in the style guide. I will locate the style guide online and send a link, but this might have to wait until the new year.

> - Measuring the performance of the database using pgbench, the default benchmarking tool

Author: Again, names of tools we consistently do not use any tagging on.

  distributed with PostgreSQL

!!! Important
    You must measure both the storage and database performance
    before putting the database into production. These results
    are extremely valuable not just in the planning phase (for example, capacity planning).
    They are also valuable in the production lifecycle, especially in emergency situations
    when you don't have time run this kind of test. Databases
    change and evolve over time, and so does the distribution of data, potentially affecting
    performance. Knowing the theoretical maximum throughput of sequential reads or
    writes is extremely useful in those situations. This is true especially in
    shared-nothing contexts, where results don't vary due to the influence of external workloads.

>     Know your system: benchmark it.

Author: Okay, I'll make it a separate paragraph. We have no standard for using bold for emphasis; the documentation standard for letting people know something is important is to use an admonition (note, important, or warning).

>     Know your system: benchmark it.

Author: ```suggestion

    Know your system: benchmark it.
```

## Encryption at rest

Encryption at rest is possible with CloudNativePG. The operator delegates that
to the underlying storage class. See the storage class for
information about this important security feature.

> ## Persistent volume claim

Author: Hm. I had checked the source doc but hadn't seen the usage in the text, only the PersistentVolumeClaim form. I will revert it.

The operator creates a persistent volume claim (PVC) for each PostgreSQL
instance, with the goal of storing the `PGDATA`. It then mounts it into each pod.

Additionally, it supports creating clusters with:

- A separate PVC on which to store PostgreSQL write-ahead log (WAL), as
  explained in [Volume for WAL](#volume-for-wal)
- Additional separate volumes reserved for PostgreSQL tablespaces, as explained

>   in [Tablespaces](tablespaces.md)

Author: All are phrases

In CloudNativePG, the volumes attached to a single PostgreSQL instance
are defined as a *PVC group*.

## Configuration via a storage class

!!! Important

>     CloudNativePG wasn't designed to work with a specific storage class.

Author: I was concerned with the use of the word "agnostic" which is jargony and might not translate to non-native English speakers or readers from different cultures. Can you come up with a better wording?

>     CloudNativePG wasn't designed to work with a specific storage class.

Reviewer: hm, good point on "agnostic". How about interoperable, interchangeable?
```suggestion
    CloudNativePG was designed to interoperate with any storage class.
```

>     CloudNativePG wasn't designed to work with a specific storage class.

Reviewer: or
```suggestion
    CloudNativePG was designed to work interchangeably with all storage classes.
```

In most cases, having `pg_wal` on the same volume where `PGDATA`
resides is fine. However, having WALs stored in a separate
volume has a few benefits:

- **I/O performance** – By storing WAL files on different storage from `PGDATA`,
  PostgreSQL can exploit parallel I/O for WAL operations (normally
  sequential writes) and for data files (tables and indexes for example), thus
  improving vertical scalability.

- **More reliability** – By reserving dedicated disk space to WAL files, you
  can be sure that exhausting space on the `PGDATA` volume
  never interferes with WAL writing. This behavior ensures that your PostgreSQL primary
  is correctly shut down.

- **Finer control** – You can define the amount of space dedicated to both
  `PGDATA` and `pg_wal`, fine tune [WAL
  configuration](https://www.postgresql.org/docs/current/wal-configuration.html)
  and checkpoints, and even use a different storage class for cost optimization.

- **Better I/O monitoring** – You can constantly monitor the load and disk usage
  on both `PGDATA` and `pg_wal`. You can also set alerts that notify you in case,
  for example, `PGDATA` requires resizing.

> !!! Seealso "Write-ahead log (WAL)"

Author: I've consistently edited this to lower case throughout our doc--it's certainly a mixed bag out there in terms of whether or not it's capitalized. But this case is a little odd because it's a reference to a topic where it's capitalized, so I'm fine if you feel strongly about it.

storage volumes and then create the related `PersistentVolume` objects for
their representation inside the Kubernetes cluster. This is also known as
*pre-provisioning* of volumes.

!!! Important
    We recommend that you avoid pre-provisioning volumes, as it
    has an effect on the high availability and self-healing capabilities
    of the operator. It breaks the fully declarative model on which
    CloudNativePG was built.

To use a pre-provisioned volume in CloudNativePG:

1. Manually create the volume outside Kubernetes.
2. Create the `PersistentVolume` object to match this volume using the
   correct parameters as required by the actual CSI driver (that is, `volumeHandle`,
   `fsType`, `storageClassName`, and so on).
3. Create the Postgres `Cluster` using, for each storage section, a coherent
   [`pvcTemplate`](storage.md#configuration-via-a-pvc-template)
   section that can help Kubernetes match the `PersistentVolume`
   and enable CloudNativePG to create the needed `PersistentVolumeClaim`.

!!! Warning
    With static provisioning, it's your responsibility to ensure that 
    Postgres pods can be correctly scheduled

>     by Kubernetes where a pre-provisioned volume exists. (This configuration is based
>     on the affinity rules of your cluster.) Make sure you check

Reviewer: ```suggestion
    by Kubernetes where a pre-provisioned volume exists. (The scheduling configuration is based
    on the affinity rules of your cluster.) Make sure you check
```

    for any pods stuck in `Pending` after you deploy the cluster. 
    If the condition persists, investigate why it's happening.

## Block storage considerations (Ceph/ Longhorn)

> Most block storage solutions in Kubernetes are suggested to have multiple replicas of a volume

Reviewer: ```suggestion
Most block storage solutions in Kubernetes recommend having multiple replicas of a volume
```
